<font size=5>DSCI-552 Homework7



#### Name: Boyu Shen
#### Github username: boyushen0701
#### USC ID: 3547352504


In [27]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.svm import LinearSVC
from imblearn.over_sampling import SMOTE
from imblearn.pipeline import Pipeline as imbpipeline
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

1. Multi-class and Multi-Label Classification Using Support Vector Machines
<br>(a) Download the Anuran Calls (MFCCs) Data Set from: https://archive.ics.uci.edu/ml/datasets/Anuran+Calls+%28MFCCs%29. 
Choose 70% of the data randomly as the training set.

In [2]:
data=pd.read_csv('../data/Anuran Calls (MFCCs)/Frogs_MFCCs.csv')
data #take a look at the data

MFCCs_ 1  MFCCs_ 2  MFCCs_ 3  MFCCs_ 4  MFCCs_ 5  MFCCs_ 6  MFCCs_ 7  \
0          1.0  0.152936 -0.105586  0.200722  0.317201  0.260764  0.100945   
1          1.0  0.171534 -0.098975  0.268425  0.338672  0.268353  0.060835   
2          1.0  0.152317 -0.082973  0.287128  0.276014  0.189867  0.008714   
3          1.0  0.224392  0.118985  0.329432  0.372088  0.361005  0.015501   
4          1.0  0.087817 -0.068345  0.306967  0.330923  0.249144  0.006884   
...        ...       ...       ...       ...       ...       ...       ...   
7190       1.0 -0.554504 -0.337717  0.035533  0.034511  0.443451  0.093889   
7191       1.0 -0.517273 -0.370574  0.030673  0.068097  0.402890  0.096628   
7192       1.0 -0.582557 -0.343237  0.029468  0.064179  0.385596  0.114905   
7193       1.0 -0.519497 -0.307553 -0.004922  0.072865  0.377131  0.086866   
7194       1.0 -0.508833 -0.324106  0.062068  0.078211  0.397188  0.094596   

      MFCCs_ 8  MFCCs_ 9  MFCCs_10  ...  MFCCs_17  MFCCs_18  MFCCs_19  \
0    -0.150063 -0.171128  0.124676  ... -0.108351 -0.077623 -0.009568   
1    -0.222475 -0.207693  0.170883  ... -0.090974 -0.056510 -0.035303   
2    -0.242234 -0.219153  0.232538  ... -0.050691 -0.023590 -0.066722   
3    -0.194347 -0.098181  0.270375  ... -0.136009 -0.177037 -0.130498   
4    -0.265423 -0.172700  0.266434  ... -0.048885 -0.053074 -0.088550   
...        ...       ...       ...  ...       ...       ...       ...   
7190 -0.100753  0.037087  0.081075  ...  0.069430  0.071001  0.021591   
7191 -0.116460  0.063727  0.089034  ...  0.061127  0.068978  0.017745   
7192 -0.103317  0.070370  0.081317  ...  0.082474  0.077771 -0.009688   
7193 -0.115799  0.056979  0.089316  ...  0.051796  0.069073  0.017963   
7194 -0.117672  0.058874  0.076180  ...  0.061455  0.072983 -0.003980   

      MFCCs_20  MFCCs_21  MFCCs_22           Family      Genus  \
0     0.057684  0.118680  0.014038  Leptodactylidae  Adenomera   
1     0.020140  0.082263  0.029056  Leptodactylidae  Adenomera   
2    -0.025083  0.099108  0.077162  Leptodactylidae  Adenomera   
3    -0.054766 -0.018691  0.023954  Leptodactylidae  Adenomera   
4    -0.031346  0.108610  0.079244  Leptodactylidae  Adenomera   
...        ...       ...       ...              ...        ...   
7190  0.052449 -0.021860 -0.079860          Hylidae     Scinax   
7191  0.046461 -0.015418 -0.101892          Hylidae     Scinax   
7192  0.027834 -0.000531 -0.080425          Hylidae     Scinax   
7193  0.041803 -0.027911 -0.096895          Hylidae     Scinax   
7194  0.031560 -0.029355 -0.087910          Hylidae     Scinax   

             Species  RecordID  
0     AdenomeraAndre         1  
1     AdenomeraAndre         1  
2     AdenomeraAndre         1  
3     AdenomeraAndre         1  
4     AdenomeraAndre         1  
...              ...       ...  
7190     ScinaxRuber        60  
7191     ScinaxRuber        60  
7192     ScinaxRuber        60  
7193     ScinaxRuber        60  
7194     ScinaxRuber        60  

[7195 rows x 26 columns]

In [3]:
from sklearn.model_selection import train_test_split

X_set=data.iloc[:,:-4]
Y_set=data.iloc[:,-4:-1] # Families, Genus, and Species are labels
X_train,X_test,Y_train,Y_test=train_test_split(X_set,Y_set,test_size=0.3,random_state=22)
X_train

MFCCs_ 1  MFCCs_ 2  MFCCs_ 3  MFCCs_ 4  MFCCs_ 5  MFCCs_ 6  MFCCs_ 7  \
2638       1.0  0.324796  0.312712  0.571277  0.117794  0.051736 -0.108748   
4029       1.0  0.343922  0.202815  0.554668  0.189612 -0.032036 -0.214515   
3359       1.0  0.157770  0.123564  0.518031  0.216295  0.048151 -0.099028   
2789       1.0  0.446230  0.293763  0.521786  0.165828  0.064217 -0.067240   
3923       1.0  0.499275  0.295859  0.555787  0.196173 -0.025651 -0.114093   
...        ...       ...       ...       ...       ...       ...       ...   
356        1.0  0.618774  0.674925  0.564547 -0.187268 -0.028974  0.509547   
7104       1.0  0.255767  0.379813  0.312270  0.283792  0.096846 -0.203001   
4908       1.0  0.069835  0.087514  0.392737  0.237009  0.258090  0.083343   
6276       1.0  0.151640  0.317315  0.288890  0.209712  0.272962  0.105791   
2933       1.0  0.251380  0.216083  0.556081  0.159532  0.098090 -0.067860   

      MFCCs_ 8  MFCCs_ 9  MFCCs_10  ...  MFCCs_13  MFCCs_14  MFCCs_15  \
2638  0.021625  0.270109  0.056149  ...  0.376238 -0.166292 -0.287838   
4029 -0.022074  0.275364  0.010725  ...  0.337136 -0.082354 -0.243919   
3359  0.022988  0.235824  0.071091  ...  0.350292  0.035715 -0.250754   
2789  0.040621  0.282114  0.011836  ...  0.423472 -0.027754 -0.253478   
3923  0.118516  0.322047  0.071258  ...  0.360658  0.008901 -0.249021   
...        ...       ...       ...  ...       ...       ...       ...   
356   0.086252 -0.266585  0.180334  ... -0.076202  0.139512  0.024253   
7104  0.075483  0.104501 -0.005680  ...  0.006804  0.011851 -0.010567   
4908 -0.312759 -0.151686  0.381647  ... -0.055604  0.335923  0.050359   
6276 -0.187052 -0.011356  0.127442  ...  0.113878 -0.072290 -0.130444   
2933  0.021326  0.271163  0.102541  ...  0.399472 -0.137104 -0.261332   

      MFCCs_16  MFCCs_17  MFCCs_18  MFCCs_19  MFCCs_20  MFCCs_21  MFCCs_22  
2638  0.093755  0.231317  0.009849 -0.128929 -0.151440  0.032300  0.211260  
4029  0.078113  0.168210 -0.040252 -0.158168 -0.178164  0.047615  0.240486  
3359 -0.020449  0.223493  0.157406 -0.009409 -0.181462 -0.101058  0.158452  
2789  0.153014  0.240175  0.059994  0.007772 -0.060229 -0.014213  0.165095  
3923 -0.006190  0.183361  0.043871 -0.072323 -0.106102  0.035438  0.190355  
...        ...       ...       ...       ...       ...       ...       ...  
356  -0.075820  0.059986  0.029501 -0.030409  0.024515  0.097357 -0.075564  
7104  0.087923  0.050572  0.004450  0.013495 -0.011554  0.027697  0.021368  
4908 -0.233119 -0.175891  0.045890  0.186029  0.206889 -0.029906 -0.285516  
6276  0.043723  0.126357  0.043502  0.005903 -0.016535 -0.078534  0.019283  
2933  0.209356  0.259244 -0.006333 -0.061217 -0.072738  0.109412  0.254989  

[5036 rows x 22 columns]

(b) Each instance has three labels: Families, Genus, and Species. Each of the labels has multiple classes. We wish to solve a multi-class and multi-label problem. One of the most important approaches to multi-label classification is to train a classifier for each label (binary relevance). We first try this approach:
<br>
<br>i. Research exact match and hamming score/ loss methods for evaluating multi-label classification and use them in evaluating the classifiers in this problem.

(1)Exact match: the percentage/accuracy of samples whose predicted set of labels exactly match with the true set of labels.
<br>(2)Hamming score: the average of correctly predicted label intersection over labels union for each sample
<br>(3)Hamming loss: the fraction of the wrong labels to the total number of labels.

The below snippet code of defining hamming score function refers from this website: https://www.linkedin.com/pulse/hamming-score-multi-label-classification-chandra-sharat

In [15]:
# reference website: https://www.linkedin.com/pulse/hamming-score-multi-label-classification-chandra-sharat
def hamming_score(y_true, y_pred):
    acc_list = []
    for i in range(y_true.shape[0]):
        set_true = set( np.where(y_true[i])[0] )
        set_pred = set( np.where(y_pred[i])[0] )
        tmp_a = None
        if len(set_true) == 0 and len(set_pred) == 0:
            tmp_a = 1
        else:
            tmp_a = len(set_true.intersection(set_pred))/float( len(set_true.union(set_pred)) )
        acc_list.append(tmp_a)
    return np.mean(acc_list)

But it seems it is for binary classification only, here it is a multiclass classification problem, so Hamming score should be like the 
average proportion of the predicted correct labels to the total number of labels over all instances(refers from "Accuracy (A)" in the paper 《A Literature Survey on Algorithms for Multi-label Learning》, by Mohammad Sorower), which is equal to 1-Hammingloss here.

ii. Train a SVM for each of the labels, using Gaussian kernels and one versus all classifiers. Determine the weight of the SVM penalty and 
the width of the Gaussian Kernel using 10 fold cross validation. You are welcome to try to solve the problem with both standardized and raw 
attributes and report the results.

In [34]:
from sklearn import preprocessing

le1 = preprocessing.LabelEncoder() # encode 4 Families, 8 Genus and 10 species with integers
Y1_train=Y_train['Family']
Y1_train=le1.fit_transform(Y1_train)
print(list(range(4)),'represent',le1.inverse_transform([0,1,2,3]),'respectively') 

[0, 1, 2, 3] represent ['Bufonidae' 'Dendrobatidae' 'Hylidae' 'Leptodactylidae'] respectively


In [35]:
le2 = preprocessing.LabelEncoder()
Y2_train=Y_train['Genus']
Y2_train=le2.fit_transform(Y2_train)
print(list(range(8)),'represent',le2.inverse_transform([0,1,2,3,4,5,6,7]),'respectively')

[0, 1, 2, 3, 4, 5, 6, 7] represent ['Adenomera' 'Ameerega' 'Dendropsophus' 'Hypsiboas' 'Leptodactylus'
 'Osteocephalus' 'Rhinella' 'Scinax'] respectively


In [36]:
le3 = preprocessing.LabelEncoder()
Y3_train=Y_train['Species']
Y3_train=le3.fit_transform(Y3_train)
print(list(range(10)),'represent',le3.inverse_transform([0,1,2,3,4,5,6,7,8,9]),'respectively')

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9] represent ['AdenomeraAndre' 'AdenomeraHylaedactylus' 'Ameeregatrivittata'
 'HylaMinuta' 'HypsiboasCinerascens' 'HypsiboasCordobae'
 'LeptodactylusFuscus' 'OsteocephalusOophagus' 'Rhinellagranulosa'
 'ScinaxRuber'] respectively


In [6]:
y1_train=Y_train['Family']
y2_train=Y_train['Genus']
y3_train=Y_train['Species']
y1_test=Y_test['Family']
y2_test=Y_test['Genus']
y3_test=Y_test['Species']

In [46]:
from sklearn import preprocessing

std=preprocessing.StandardScaler()
std.fit(X_train)
X_train_std=std.transform(X_train)
X_test_std=std.transform(X_test)

In [56]:
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV

lambdas=[]
log_l=range(-3,5)
for i in log_l:
    lambdas.append(10**i)
gammas=np.arange(0.1,2.1,0.2)

svc=SVC(kernel='rbf', random_state=22, max_iter = 100, decision_function_shape='ovr') # train Gaussian kernels and one versus all classifiers
parameters={'C':lambdas,'gamma':gammas}

gs1 = GridSearchCV(svc, parameters, cv=10 ,n_jobs=-1).fit(X_train, y1_train)
c1=gs1.best_params_['C']
gamma1=gs1.best_params_['gamma']
print('For SVC predicting label "Family",','the best c is',c1,'and the best gamma is',gamma1)

For SVC predicting label "Family", the best c is 10 and the best gamma is 1.7000000000000004


E:\usc\anaconda\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=100).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


In [49]:
gs2 = GridSearchCV(svc, parameters, cv=10 ,n_jobs=-1).fit(X_train, y2_train)
c2=gs2.best_params_['C']
gamma2=gs2.best_params_['gamma']
print('For SVC predicting label "Genus",','the best c is',c2,'and the best gamma is',gamma2)

For SVC predicting label "Genus", the best c is 10 and the best gamma is 1.9000000000000004


E:\usc\anaconda\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=100).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


In [50]:
gs3 = GridSearchCV(svc, parameters, cv=10 ,n_jobs=-1).fit(X_train, y3_train)
c3=gs3.best_params_['C']
gamma3=gs3.best_params_['gamma']
print('For SVC predicting label "Species",','the best c is',c3,'and the best gamma is',gamma3)

For SVC predicting label "Species", the best c is 10 and the best gamma is 1.7000000000000004


E:\usc\anaconda\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=100).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


In [58]:
svc1=SVC(C=c1,gamma=gamma1,kernel='rbf',random_state=22,max_iter=100,decision_function_shape='ovr')
svc1.fit(X_train, y1_train) # re-train the model with whole training data
y1_pred=svc1.predict(X_test)
print(y1_pred)

['Leptodactylidae' 'Hylidae' 'Leptodactylidae' ... 'Hylidae'
 'Dendrobatidae' 'Hylidae']


E:\usc\anaconda\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=100).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


In [59]:
svc2=SVC(C=c2,gamma=gamma2,kernel='rbf',random_state=22,max_iter=100,decision_function_shape='ovr')
svc2.fit(X_train, y2_train) # re-train the model with whole training data
y2_pred=svc2.predict(X_test)
print(y2_pred)

['Adenomera' 'Hypsiboas' 'Adenomera' ... 'Dendropsophus' 'Ameerega'
 'Hypsiboas']


E:\usc\anaconda\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=100).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


In [60]:
svc3=SVC(C=c3,gamma=gamma3,kernel='rbf',random_state=22,max_iter=100,decision_function_shape='ovr')
svc3.fit(X_train, y3_train) # re-train the model with whole training data
y3_pred=svc3.predict(X_test)
print(y3_pred)

['AdenomeraHylaedactylus' 'HypsiboasCordobae' 'AdenomeraHylaedactylus' ...
 'HylaMinuta' 'Ameeregatrivittata' 'HypsiboasCinerascens']


E:\usc\anaconda\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=100).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


In [63]:
zipped=list(zip(y1_pred,y2_pred,y3_pred))
print(zipped) # multilabel predictions for test set

[('Leptodactylidae', 'Adenomera', 'AdenomeraHylaedactylus'), ('Hylidae', 'Hypsiboas', 'HypsiboasCordobae'), ('Leptodactylidae', 'Adenomera', 'AdenomeraHylaedactylus'), ('Leptodactylidae', 'Adenomera', 'AdenomeraHylaedactylus'), ('Hylidae', 'Dendropsophus', 'HylaMinuta'), ('Hylidae', 'Hypsiboas', 'HypsiboasCordobae'), ('Leptodactylidae', 'Leptodactylus', 'LeptodactylusFuscus'), ('Leptodactylidae', 'Adenomera', 'AdenomeraAndre'), ('Leptodactylidae', 'Adenomera', 'AdenomeraHylaedactylus'), ('Hylidae', 'Dendropsophus', 'HylaMinuta'), ('Dendrobatidae', 'Ameerega', 'Ameeregatrivittata'), ('Leptodactylidae', 'Adenomera', 'AdenomeraHylaedactylus'), ('Leptodactylidae', 'Adenomera', 'AdenomeraHylaedactylus'), ('Leptodactylidae', 'Adenomera', 'AdenomeraAndre'), ('Leptodactylidae', 'Adenomera', 'AdenomeraHylaedactylus'), ('Leptodactylidae', 'Adenomera', 'AdenomeraAndre'), ('Leptodactylidae', 'Adenomera', 'AdenomeraHylaedactylus'), ('Hylidae', 'Hypsiboas', 'HypsiboasCordobae'), ('Dendrobatidae', 'A

In [9]:
def exact_match(Y1,Y2):
    n1=len(Y1)
    n2=len(Y1[0])
    count=0
    for i in range(n1):
        exact=True
        for j in range(n2):
            if Y1[i][j]!=Y2[i][j]:
                exact=False
        if exact==True:
            count+=1
    return count/n1

def hamming_loss(Y1,Y2):
    n1=len(Y1)
    n2=len(Y1[0])
    count=0
    for i in range(n1):
        for j in range(n2):
            if Y1[i][j]==Y2[i][j]:
                count+=1
    return 1-count/(n1*n2)

In [76]:
actual=[]
for index,row in Y_test.iterrows():
    actual.append(row.tolist())

print("Before standardized:")
print('The Exact match accuracy is',exact_match(zipped,actual))
print('The Hamming loss is',hamming_loss(zipped,actual))
print('The Hamming score is',1-hamming_loss(zipped,actual))

Before standardized:
The Exact match accuracy is 0.9851783232978231
The Hamming loss is 0.008954763007565258
The Hamming score is 0.9910452369924347


In [51]:
# try with standardized data
gs4 = GridSearchCV(svc, parameters, cv=10 ,n_jobs=-1).fit(X_train_std, y1_train)
c4=gs4.best_params_['C']
gamma4=gs4.best_params_['gamma']
print('For SVC predicting label "Family" after standardizing,','the best c is',c4,'and the best gamma is',gamma4)

For SVC predicting label "Family" after standardizing, the best c is 10 and the best gamma is 0.1


E:\usc\anaconda\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=100).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


In [53]:
gs5 = GridSearchCV(svc, parameters, cv=10 ,n_jobs=-1).fit(X_train_std, y2_train)
c5=gs5.best_params_['C']
gamma5=gs5.best_params_['gamma']
print('For SVC predicting label "Genus" after standardizing,','the best c is',c5,'and the best gamma is',gamma5)

For SVC predicting label "Genus" after standardizing, the best c is 10 and the best gamma is 0.1


E:\usc\anaconda\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=100).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


In [54]:
gs6 = GridSearchCV(svc, parameters, cv=10 ,n_jobs=-1).fit(X_train_std, y3_train)
c6=gs6.best_params_['C']
gamma6=gs6.best_params_['gamma']
print('For SVC predicting label "Species" after standardizing,','the best c is',c6,'and the best gamma is',gamma6)

For SVC predicting label "Species" after standardizing, the best c is 10 and the best gamma is 0.1


E:\usc\anaconda\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=100).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


In [94]:
svc4=SVC(C=c4,gamma=gamma4,kernel='rbf',random_state=22,max_iter=100,decision_function_shape='ovr')
svc4.fit(X_train_std, y1_train) # re-train the model with whole standardized training data
y4_pred=svc4.predict(X_test_std)
print(y4_pred)

['Leptodactylidae' 'Hylidae' 'Leptodactylidae' ... 'Hylidae'
 'Dendrobatidae' 'Hylidae']


E:\usc\anaconda\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=100).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


In [95]:
svc5=SVC(C=c5,gamma=gamma5,kernel='rbf',random_state=22,max_iter=100,decision_function_shape='ovr')
svc5.fit(X_train_std, y2_train) # re-train the model with whole standardized training data
y5_pred=svc5.predict(X_test_std)
print(y5_pred)

E:\usc\anaconda\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=100).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


['Adenomera' 'Hypsiboas' 'Adenomera' ... 'Dendropsophus' 'Ameerega'
 'Hypsiboas']


In [96]:
svc6=SVC(C=c6,gamma=gamma6,kernel='rbf',random_state=22,max_iter=100,decision_function_shape='ovr')
svc6.fit(X_train_std, y3_train) # re-train the model with whole standardized training data
y6_pred=svc6.predict(X_test_std)
print(y6_pred)

E:\usc\anaconda\lib\site-packages\sklearn\svm\_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=100).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


['AdenomeraHylaedactylus' 'HypsiboasCordobae' 'AdenomeraHylaedactylus' ...
 'HylaMinuta' 'Ameeregatrivittata' 'HypsiboasCinerascens']


In [97]:
zipped1=list(zip(y4_pred,y5_pred,y6_pred))
print("After standardized:")
print('The Exact match accuracy is',exact_match(zipped1,actual))
print('The Hamming loss is',hamming_loss(zipped1,actual))
print('The Hamming score is',1-hamming_loss(zipped1,actual))

After standardized:
The Exact match accuracy is 0.9805465493283928
The Hamming loss is 0.012968967114404872
The Hamming score is 0.9870310328855951


It seems that the results before standardized is slightly better than the results after standardized.

iii. Repeat 1(b)ii with L1-penalized SVMs. Remember to standardize the attributes. Determine the weight of the SVM penalty using 10 fold 
cross validation.

In [85]:
from sklearn.svm import LinearSVC

lsvc=LinearSVC(random_state=22, penalty='l1',dual=False ,multi_class='ovr',max_iter=5000) # n_samples > n_features, linear kernel
parameters={'C':lambdas}

GS1 = GridSearchCV(lsvc, parameters, cv=10 ,n_jobs=-1).fit(X_train_std, y1_train) # use standardized data
c1_l=GS1.best_params_['C']
print('For LinearSVC predicting label "Family",','the best c is',c1_l)

For LinearSVC predicting label "Family", the best c is 1


In [86]:
GS2 = GridSearchCV(lsvc, parameters, cv=10 ,n_jobs=-1).fit(X_train_std, y2_train) # use standardized data
c2_l=GS2.best_params_['C']
print('For LinearSVC predicting label "Genus",','the best c is',c2_l)

For LinearSVC predicting label "Genus", the best c is 10


In [87]:
GS3 = GridSearchCV(lsvc, parameters, cv=10 ,n_jobs=-1).fit(X_train_std, y3_train) # use standardized data
c3_l=GS3.best_params_['C']
print('For LinearSVC predicting label "Species",','the best c is',c3_l)

For LinearSVC predicting label "Species", the best c is 100


In [98]:
lsvc1=LinearSVC(C=c1_l,random_state=22,penalty='l1',dual=False ,multi_class='ovr',max_iter=5000)
lsvc1.fit(X_train_std, y1_train) # re-train the model with whole standardized training data
y1_linear_pred=lsvc1.predict(X_test_std)
print(y1_linear_pred)

['Leptodactylidae' 'Hylidae' 'Leptodactylidae' ... 'Hylidae'
 'Dendrobatidae' 'Hylidae']


In [99]:
lsvc2=LinearSVC(C=c2_l,random_state=22,penalty='l1',dual=False ,multi_class='ovr',max_iter=5000)
lsvc2.fit(X_train_std, y2_train) # re-train the model with whole standardized training data
y2_linear_pred=lsvc2.predict(X_test_std)
print(y2_linear_pred)

['Adenomera' 'Hypsiboas' 'Adenomera' ... 'Dendropsophus' 'Ameerega'
 'Hypsiboas']


In [100]:
lsvc3=LinearSVC(C=c3_l,random_state=22,penalty='l1',dual=False ,multi_class='ovr',max_iter=5000)
lsvc3.fit(X_train_std, y3_train) # re-train the model with whole standardized training data
y3_linear_pred=lsvc3.predict(X_test_std)
print(y3_linear_pred)

['AdenomeraHylaedactylus' 'HypsiboasCordobae' 'AdenomeraHylaedactylus' ...
 'HylaMinuta' 'Ameeregatrivittata' 'HypsiboasCinerascens']


In [101]:
zipped2=list(zip(y1_linear_pred,y2_linear_pred,y3_linear_pred))
print("For L1-penalized SVM:")
print('The Exact match accuracy is',exact_match(zipped2,actual))
print('The Hamming loss is',hamming_loss(zipped2,actual))
print('The Hamming score is',1-hamming_loss(zipped2,actual))

For L1-penalized SVM:
The Exact match accuracy is 0.9129226493747105
The Hamming loss is 0.054191755442334455
The Hamming score is 0.9458082445576655


iv. Repeat 1(b)iii by using SMOTE or any other method you know to remedy class imbalance. Report your conclusions about the classifiers you 
trained.

In [107]:
# compute the number of instances for each class of each label
dict1={}
for stuff in y1_train:
    if stuff not in dict1:
        dict1[stuff]=1
    else:
        dict1[stuff]+=1
print(dict1)
dict2={}
for stuff in y2_train:
    if stuff not in dict2:
        dict2[stuff]=1
    else:
        dict2[stuff]+=1
print(dict2)
dict3={}
for stuff in y3_train:
    if stuff not in dict3:
        dict3[stuff]=1
    else:
        dict3[stuff]+=1
print(dict3)

{'Leptodactylidae': 3092, 'Hylidae': 1517, 'Dendrobatidae': 379, 'Bufonidae': 48}
{'Adenomera': 2911, 'Hypsiboas': 1115, 'Scinax': 105, 'Osteocephalus': 87, 'Ameerega': 379, 'Dendropsophus': 210, 'Leptodactylus': 181, 'Rhinella': 48}
{'AdenomeraHylaedactylus': 2445, 'HypsiboasCordobae': 773, 'ScinaxRuber': 105, 'OsteocephalusOophagus': 87, 'Ameeregatrivittata': 379, 'AdenomeraAndre': 466, 'HylaMinuta': 210, 'HypsiboasCinerascens': 342, 'LeptodactylusFuscus': 181, 'Rhinellagranulosa': 48}


We can observe that there is obvious class imbalance problems.

In [108]:
from imblearn.over_sampling import SMOTE
from imblearn.pipeline import Pipeline as imbpipeline

pipeline = imbpipeline(steps = [['smote', SMOTE(random_state=42)],
                                ['classifier',  LinearSVC(random_state=42,penalty='l1',dual=False ,multi_class='ovr',max_iter=5000)]])
para={'classifier__C':lambdas} 
GS = GridSearchCV(estimator=pipeline, param_grid=para, n_jobs=-1, cv=10) # include the SMOTE within the cross-validation process
GS.fit(X_train_std,y1_train)
smote_c1=GS.best_params_
smote_best_score1=GS.best_score_

print('After SMOTE, the best c for label "Family" is',smote_c1,',the corresponding validation accuracy is',smote_best_score1)

After SMOTE, the best c for label "Family" is {'classifier__C': 100} ,the corresponding validation accuracy is 0.9249439868724163


In [113]:
GS = GridSearchCV(estimator=pipeline, param_grid=para, n_jobs=-1, cv=10)
GS.fit(X_train_std,y2_train)
smote_c2=GS.best_params_
smote_best_score2=GS.best_score_

print('After SMOTE, the best c for label "Genus" is',smote_c2,',the corresponding validation accuracy is',smote_best_score2)

After SMOTE, the best c for label "Genus" is {'classifier__C': 10} ,the corresponding validation accuracy is 0.9171987913787121


In [114]:
GS = GridSearchCV(estimator=pipeline, param_grid=para, n_jobs=-1, cv=10)
GS.fit(X_train_std,y3_train)
smote_c3=GS.best_params_
smote_best_score3=GS.best_score_

print('After SMOTE, the best c for label "Species" is',smote_c3,',the corresponding validation accuracy is',smote_best_score3)

After SMOTE, the best c for label "Species" is {'classifier__C': 1} ,the corresponding validation accuracy is 0.9567121083025654


In [116]:
smote_c1=smote_c1['classifier__C']
smo=SMOTE(random_state=42)
x_smo1,y_smo1=smo.fit_resample(X_train_std,y1_train) # reconstruct the same smote samples as in cross-validation process
smo_svc1=LinearSVC(C=smote_c1,random_state=42,penalty='l1',dual=False ,multi_class='ovr',max_iter=5000)
smo_svc1.fit(x_smo1, y_smo1) # re-train the model with smote training data
y1_smote_pred=smo_svc1.predict(X_test_std)
print(y1_smote_pred)

['Leptodactylidae' 'Hylidae' 'Leptodactylidae' ... 'Hylidae'
 'Dendrobatidae' 'Hylidae']


In [117]:
smote_c2=smote_c2['classifier__C']
smo=SMOTE(random_state=42)
x_smo2,y_smo2=smo.fit_resample(X_train_std,y2_train) # reconstruct the same smote samples as in cross-validation process
smo_svc2=LinearSVC(C=smote_c2,random_state=42,penalty='l1',dual=False ,multi_class='ovr',max_iter=5000)
smo_svc2.fit(x_smo2, y_smo2) # re-train the model with smote training data
y2_smote_pred=smo_svc2.predict(X_test_std)
print(y2_smote_pred)

['Adenomera' 'Hypsiboas' 'Adenomera' ... 'Dendropsophus' 'Ameerega'
 'Hypsiboas']


In [118]:
smote_c3=smote_c3['classifier__C']
smo=SMOTE(random_state=42)
x_smo3,y_smo3=smo.fit_resample(X_train_std,y3_train) # reconstruct the same smote samples as in cross-validation process
smo_svc3=LinearSVC(C=smote_c3,random_state=42,penalty='l1',dual=False ,multi_class='ovr',max_iter=5000)
smo_svc3.fit(x_smo3, y_smo3) # re-train the model with smote training data
y3_smote_pred=smo_svc3.predict(X_test_std)
print(y3_smote_pred)

['AdenomeraHylaedactylus' 'HypsiboasCordobae' 'AdenomeraHylaedactylus' ...
 'HylaMinuta' 'Ameeregatrivittata' 'HypsiboasCinerascens']


In [119]:
zipped3=list(zip(y1_smote_pred,y2_smote_pred,y3_smote_pred))
print("After SMOTE, for L1-penalized SVM:")
print('The Exact match accuracy is',exact_match(zipped3,actual))
print('The Hamming loss is',hamming_loss(zipped3,actual))
print('The Hamming score is',1-hamming_loss(zipped3,actual))

After SMOTE, for L1-penalized SVM:
The Exact match accuracy is 0.8624363131079203
The Hamming loss is 0.07302763625135089
The Hamming score is 0.9269723637486491


2. K-Means Clustering on a Multi-Class and Multi-Label Data Set
<br>Monte-Carlo Simulation: Perform the following procedures 50 times, and report the average and standard deviation of the 50 Hamming Distances that you calculate.

(a) Use k-means clustering on the whole Anuran Calls (MFCCs) Data Set (do not split the data into train and test, as we are not performing 
supervised learning in this exercise). Choose k ∈ {1,2,...,50} automatically based on one of the methods provided in the slides (CH or Gap 
Statistics or scree plots or Silhouettes) or any other method you know.
<br>(b) In each cluster, determine which family is the majority by reading the true labels. Repeat for genus and species.
<br>(c) Now for each cluster you have a majority label triplet (family, genus, species). Calculate the average Hamming distance, Hamming score, and Hamming loss between the true labels and the labels assigned by clusters.

In [11]:
y1_all=Y_set['Family']
y2_all=Y_set['Genus']
y3_all=Y_set['Species']

actual_all=[]
for index,row in Y_set.iterrows():
    actual_all.append(row.tolist())

def hamming_distance(Y1,Y2): # the average hamming distance in each simulation
    n1=len(Y1)
    n2=len(Y1[0])
    dis_sum=0
    for num1 in range(n1):
        for num2 in range(n2):
            if Y1[num1][num2]!=Y2[num1][num2]:
                dis_sum+=1
    return dis_sum/n1

In [19]:
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

X_set # the whole feature set
rs=range(50)
Ks=range(2,51)
distances=[]
losses=[]
scores=[]
for i in rs: # for each simulation
    family_assign={}
    genus_assign={}
    species_assign={}
    best_k=[]
    best_silhouette=[]
    cluster_label=[]
    bestk=1
    bestscore=-1 # silhouette score for k=1
    for k in Ks: # for each k
        cluster=KMeans(n_clusters=k,random_state=i) # the ith simulation
        cluster.fit(X_set)
        labels=cluster.labels_
        score=silhouette_score(X_set,labels)
        if score>bestscore:
            bestk=k
            bestscore=score
            cluster_label=labels
    best_k.append(bestk)
    best_silhouette.append(bestscore)
    for j in range(bestk): # for each cluster with the best k
        families=[]
        genus=[]
        species=[]
        for l in range(len(cluster_label)):
            if cluster_label[l]==j:
                families.append(y1_all[l])
                genus.append(y2_all[l])
                species.append(y3_all[l])
        dic1={}
        dic2={}
        dic3={}
        for fam in families:
            if fam not in dic1:
                dic1[fam]=families.count(fam)
        family_assign[j]=sorted([(v1,k1) for k1,v1 in dic1.items()],reverse=True)[0][1]
        for gen in genus:
            if gen not in dic2:
                dic2[gen]=genus.count(gen)
        genus_assign[j]=sorted([(v2,k2) for k2,v2 in dic2.items()],reverse=True)[0][1]
        for spe in species:
            if spe not in dic3:
                dic3[spe]=species.count(spe)
        species_assign[j]=sorted([(v3,k3) for k3,v3 in dic3.items()],reverse=True)[0][1]
    family_pred=[]
    genus_pred=[]
    species_pred=[]
    for lab in cluster_label:
        for q in range(bestk):
            if lab==q:
                family_pred.append(family_assign[lab])
                genus_pred.append(genus_assign[lab])
                species_pred.append(species_assign[lab])
    zipped_c=list(zip(family_pred,genus_pred,species_pred)) # predicted labels for each instance
    hammingloss=hamming_loss(zipped_c,actual_all)
    hammingscore=1-hammingloss
    hammingdistance=hamming_distance(zipped_c,actual_all)
    losses.append(hammingloss)
    scores.append(hammingscore)
    distances.append(hammingdistance)
    print('Simulation',i+1,'completed')
    print('the best k is',bestk,', and the best silhouette score is',bestscore)
    print('the predicted Family label for each cluster is',family_assign)
    print('the predicted Genus label for each cluster is',genus_assign)
    print('the predicted Species label for each cluster is',species_assign)
    print('the average Hamming distance in this simulation is',hammingdistance)
    print('the Hamming score in this simulation is',hammingscore)
    print('the Hamming loss in this simulation is',hammingloss)
    

Simulation 1 completed
the best k is 4 , and the best silhouette score is 0.3788646212055977
the predicted Family label for each cluster is {0: 'Leptodactylidae', 1: 'Dendrobatidae', 2: 'Hylidae', 3: 'Hylidae'}
the predicted Genus label for each cluster is {0: 'Adenomera', 1: 'Ameerega', 2: 'Hypsiboas', 3: 'Hypsiboas'}
the predicted Species label for each cluster is {0: 'AdenomeraHylaedactylus', 1: 'Ameeregatrivittata', 2: 'HypsiboasCinerascens', 3: 'HypsiboasCordobae'}
the average Hamming distance in this simulation is 0.6653231410701876
the Hamming score in this simulation is 0.7782256196432707
the Hamming loss in this simulation is 0.22177438035672925
Simulation 2 completed
the best k is 4 , and the best silhouette score is 0.37875093433052953
the predicted Family label for each cluster is {0: 'Hylidae', 1: 'Dendrobatidae', 2: 'Leptodactylidae', 3: 'Hylidae'}
the predicted Genus label for each cluster is {0: 'Hypsiboas', 1: 'Ameerega', 2: 'Adenomera', 3: 'Hypsiboas'}
the predicted S

Simulation 14 completed
the best k is 4 , and the best silhouette score is 0.37863353074850936
the predicted Family label for each cluster is {0: 'Hylidae', 1: 'Hylidae', 2: 'Leptodactylidae', 3: 'Dendrobatidae'}
the predicted Genus label for each cluster is {0: 'Hypsiboas', 1: 'Hypsiboas', 2: 'Adenomera', 3: 'Ameerega'}
the predicted Species label for each cluster is {0: 'HypsiboasCordobae', 1: 'HypsiboasCinerascens', 2: 'AdenomeraHylaedactylus', 3: 'Ameeregatrivittata'}
the average Hamming distance in this simulation is 0.6674079221681724
the Hamming score in this simulation is 0.7775306926106093
the Hamming loss in this simulation is 0.22246930738939075
Simulation 15 completed
the best k is 4 , and the best silhouette score is 0.37875093433052953
the predicted Family label for each cluster is {0: 'Leptodactylidae', 1: 'Hylidae', 2: 'Dendrobatidae', 3: 'Hylidae'}
the predicted Genus label for each cluster is {0: 'Adenomera', 1: 'Hypsiboas', 2: 'Ameerega', 3: 'Hypsiboas'}
the predicte

Simulation 27 completed
the best k is 4 , and the best silhouette score is 0.37886255335886554
the predicted Family label for each cluster is {0: 'Hylidae', 1: 'Hylidae', 2: 'Leptodactylidae', 3: 'Dendrobatidae'}
the predicted Genus label for each cluster is {0: 'Hypsiboas', 1: 'Hypsiboas', 2: 'Adenomera', 3: 'Ameerega'}
the predicted Species label for each cluster is {0: 'HypsiboasCinerascens', 1: 'HypsiboasCordobae', 2: 'AdenomeraHylaedactylus', 3: 'Ameeregatrivittata'}
the average Hamming distance in this simulation is 0.66726893676164
the Hamming score in this simulation is 0.7775770210794534
the Hamming loss in this simulation is 0.22242297892054663
Simulation 28 completed
the best k is 4 , and the best silhouette score is 0.37863353074850936
the predicted Family label for each cluster is {0: 'Dendrobatidae', 1: 'Leptodactylidae', 2: 'Hylidae', 3: 'Hylidae'}
the predicted Genus label for each cluster is {0: 'Ameerega', 1: 'Adenomera', 2: 'Hypsiboas', 3: 'Hypsiboas'}
the predicted 

Simulation 40 completed
the best k is 4 , and the best silhouette score is 0.37875093433052953
the predicted Family label for each cluster is {0: 'Hylidae', 1: 'Hylidae', 2: 'Leptodactylidae', 3: 'Dendrobatidae'}
the predicted Genus label for each cluster is {0: 'Hypsiboas', 1: 'Hypsiboas', 2: 'Adenomera', 3: 'Ameerega'}
the predicted Species label for each cluster is {0: 'HypsiboasCinerascens', 1: 'HypsiboasCordobae', 2: 'AdenomeraHylaedactylus', 3: 'Ameeregatrivittata'}
the average Hamming distance in this simulation is 0.66726893676164
the Hamming score in this simulation is 0.7775770210794534
the Hamming loss in this simulation is 0.22242297892054663
Simulation 41 completed
the best k is 4 , and the best silhouette score is 0.37875093433052953
the predicted Family label for each cluster is {0: 'Leptodactylidae', 1: 'Hylidae', 2: 'Hylidae', 3: 'Dendrobatidae'}
the predicted Genus label for each cluster is {0: 'Adenomera', 1: 'Hypsiboas', 2: 'Hypsiboas', 3: 'Ameerega'}
the predicted 

In [25]:
dis_avg=sum(distances)/len(distances)
dis_std=np.std(distances)
loss_avg=sum(losses)/len(losses)
loss_std=np.std(losses)
score_avg=sum(scores)/len(scores)
score_std=np.std(scores)

In [26]:
stats=[[dis_avg,dis_std],[score_avg,score_std],[loss_avg,loss_std]]
report=pd.DataFrame(data=stats,index=['Hamming Distance','Hamming score','Hamming loss'],columns=['average','standard deviation'])
report

average  standard deviation
Hamming Distance  0.671169             0.03102
Hamming score     0.776277             0.01034
Hamming loss      0.223723             0.01034

3. ISLR 12.6.2

2. Suppose that we have four observations, for which we compute a dissimilarity matrix, given by
<br>&emsp; &ensp;  0.3 &nbsp; &nbsp; 0.4 &nbsp; 0.7
<br>0.3 &emsp; &ensp; &ensp; 0.5 &ensp; 0.8
<br>0.4 &nbsp; 0.5 &emsp; &emsp; &nbsp; 0.45
<br>0.7 &nbsp; 0.8 &nbsp; 0.45 &emsp;

For instance, the dissimilarity between the first and second observations is 0.3, and the dissimilarity between the second and fourth
observations is 0.8.

(a) On the basis of this dissimilarity matrix, sketch the dendrogram that results from hierarchically clustering these four observations
using complete linkage. Be sure to indicate on the plot the height at which each fusion occurs, as well as the observations
corresponding to each leaf in the dendrogram.
<br>(b) Repeat (a), this time using single linkage clustering.
<br>(c) Suppose that we cut the dendrogram obtained in (a) such that two clusters result. Which observations are in each cluster?
<br>(d) Suppose that we cut the dendrogram obtained in (b) such that two clusters result. Which observations are in each cluster?
<br>(e) It is mentioned in the chapter that at each fusion in the dendrogram, the position of the two clusters being fused can be
swapped without changing the meaning of the dendrogram. Draw a dendrogram that is equivalent to the dendrogram in (a), for which two or more 
of the leaves are repositioned, but for which the meaning of the dendrogram is the same.

Answers for these questions are in the picture attached with this homework.